### PR TITLE
SY-875 - Add warning when console and server versions mismatch

### PR DESCRIPTION
**Pull request by emilbon99**
# Feature Pull Request Template

## Key Information

- **Linear Issue**: [SY-875](https://linear.app/synnax/issue/SY-875/[console]-check-version-against-server)

## Description

Implements a version checking system in the TypeScript client that warns when there is a mismatch between the **major** or **minor** versions of the client and server.

Also makes changes to Pluto and the Console that warn with a notification when there is a version mismatch.
 
## Basic Readiness Checklist

- [x] I have performed a self-review of my code.
- [x] I have added relevant tests to cover the changes to CI.
- [x] I have updated user facing documentation accordingly.
- [x] I have verified code coverage targets are met.

## Migrations

- [ ] Console - I have ensured that previous versions of stored data structures are 
properly migrated to new formats.
- [ ] Server - I have ensured that previous versions of stored data structures are 
properly migrated to new formats.

##  Additional Notes
- [ ] These changes deal with concurrency.
- [ ] These changes affect UI.

## Manual QA Additions

- [ ] I have updated the [Release Candidate](/.github/PULL_REQUEST_TEMPLATE/rc.md) template
with necessary manual QA steps to test my change.

## Breaking Changes

Please list any breaking changes to public or internal packages.

## Reviwer Checklist
- [ ] Sufficient test coverage of new additions.
- [ ] Verified all steps in the Readiness checklists.
- [ ] UI changes have been tested.
- [ ] Style and formatting is consistent.
- [ ] Reviewed any relevant changes to concurrent code for safety. 
- [ ] Sufficient comments and clarity of code.
- [ ] Any additional documentation necessary for feature is provided and has been reviewed for clarity.

### CODE REVIEW

**Comment by Lham42** [comment]
Update title and pr description

**Comment by Lham42** [comment]
Add to the pr description - do we check the just major and minor or also the patch version?

**Comment by emilbon99** [comment]
@Lham42 Updated

**Review by Lham42 — CHANGES_REQUESTED** [review]

File: client/ts/src/connection/checker.ts [hunk 1]
```diff
@@ -77,11 +88,54 @@ export class Checker {
...
+        {},
+        z.object({}),
+        responseZ,
+      );
       if (err != null) throw err;
+      if (res.nodeVersion == null) {
+        this._state.clientServerCompatible = false;
+        if (!this.versionWarned)
+          console.warn(`
+          Synnax cluster node version is too old for client version ${this.clientVersion}.
```
Comment (pjdotson) [inline client/ts/src/connection/checker.ts:102]: Can you consolidate these error statements into one statement somewhere that accepts arguments for old / new and the versions? It's pretty hard to follow the logic of so many nested if-else statements and that would make it easier.

File: client/ts/src/connection/connection.spec.ts [hunk 2]
```diff
@@ -23,8 +24,35 @@ describe("connectivity", () => {
...
+        username: "synnax",
+        password: "seldon",
+      });
+      transport.use(client.middleware());
+      const connectivity = new Checker(transport.unary, undefined, __VERSION__);
+      const state = await connectivity.check();
+      expect(state.clientServerCompatible).toBe(true);
+      expect(state.clientVersion).toBe(__VERSION__);
+    });
+    it("should adjust state if the server is too old", async () => {
```
Comment (pjdotson) [inline client/ts/src/connection/connection.spec.ts:45]: Add a test for where the server is too new as well.

File: pluto/src/synnax/Provider.tsx [hunk 3]
```diff
@@ -88,20 +83,50 @@ export const Provider = Aether.wrap<ProviderProps>(
...
+        Compatibility issues may arise. 
+        `;
+        else if (connectivity.nodeVersion != null)
+          description = `Cluster version ${connectivity.nodeVersion} is older than client version ${connectivity.clientVersion}.
+        Compatibility issues may arise. 
+        `;
+        else
+          description = `Cluster version is older than client version ${connectivity.clientVersion}. Compatibility issues may arise. `;
+
+        add({
```
Comment (pjdotson) [inline pluto/src/synnax/Provider.tsx:117]: Rename this `addStatus` to match the rest of the codebase.

File: pluto/src/synnax/Provider.tsx [hunk 4]
```diff
@@ -88,20 +83,50 @@ export const Provider = Aether.wrap<ProviderProps>(
...
+      if (connectivity.status === "connected" && !connectivity.clientServerCompatible) {
+        const oldServer =
+          connectivity.nodeVersion == null ||
+          migrate.semVerOlder(connectivity.nodeVersion, connectivity.clientVersion);
+
+        let description: string;
+        if (!oldServer)
+          description = `
+        Cluster version ${connectivity.nodeVersion} is newer than client version ${connectivity.clientVersion}.
+        Compatibility issues may arise.
```
Comment (pjdotson) [inline pluto/src/synnax/Provider.tsx:108]: Personally I think we should put long strings on one line, but I don't know if there is a convention followed in the rest of the codebase.

File: pluto/src/synnax/Provider.tsx [hunk 5]
```diff
@@ -88,20 +83,50 @@ export const Provider = Aether.wrap<ProviderProps>(
...
+          description = `
+        Cluster version ${connectivity.nodeVersion} is newer than client version ${connectivity.clientVersion}.
+        Compatibility issues may arise. 
+        `;
+        else if (connectivity.nodeVersion != null)
+          description = `Cluster version ${connectivity.nodeVersion} is older than client version ${connectivity.clientVersion}.
+        Compatibility issues may arise. 
+        `;
+        else
+          description = `Cluster version is older than client version ${connectivity.clientVersion}. Compatibility issues may arise. `;
```
Comment (pjdotson) [inline pluto/src/synnax/Provider.tsx:115]: You could probably also make these strings a function that you pass older / newer / versions into, but I think this is easier to read than the client checker so you don't have to.

File: docs/site/src/pages/reference/typescript-client/troubleshooting.mdx [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+```
+Synnax cluster node version is too old for client version x.x.x.
+This may cause compatibility issues. We recommend upgrading the cluster.
+```
+</div>
+
+This warning message indicates that the the version of the Synnax TypeScript client
+you are using is newer than the version of the Synnax cluster you are connecting to. To
+resolve this issue, either [upgrade the Synnax cluster](/reference/cluster/installation) 
+or downgrade to a matching version of the client.
```
Comment (pjdotson) [inline docs/site/src/pages/reference/typescript-client/troubleshooting.mdx:28]: End this line with a colon.

File: docs/site/src/pages/reference/typescript-client/troubleshooting.mdx [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+<div class="warning">
+```
+Synnax cluster node version is too new client version x.x.x.
+This may cause compatibility issues. We recommend upgrading the client.
+```
+</div>
+
+This warning message indicates that the the version of the Synnax TypeScript client
+you are using is older than the version of the Synnax cluster you are connecting to. To
+resolve this issue, we recommend upgrading the Synnax TypeScript client.
```
Comment (pjdotson) [inline docs/site/src/pages/reference/typescript-client/troubleshooting.mdx:47]: End this line with a colon.

File: console/src/hardware/device/external.ts [hunk 8]
```diff
@@ -7,5 +7,9 @@
 // License, use of this software will be governed by the Apache License, Version 2.0,
 // included in the file licenses/APL.txt.
 
+import { notificationAdapter } from "@/version/Updater";
```
Comment (pjdotson) [inline console/src/hardware/device/external.ts:10]: You should be import `Version` from "@/version" instead of importing one file in the package.

File: console/src/cluster/notification.tsx [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+
+import {
+  NotificationAdapter,
+  SugaredNotification,
+} from "@/notifications/Notifications";
+import { OpenUpdateDialogAction } from "@/version/Updater";
+
+export const versionOutdatedAdapter: NotificationAdapter = (status) => {
+  if (status.data == null) return null;
+  if (status.data.type !== "serverVersionMismatch") return null;
```
Comment (pjdotson) [inline console/src/cluster/notification.tsx:20]: Could we change the string `"serverVersionMismatch"` to a `const` that gets exported from Pluto and imported here?

**Comment by Lham42** [comment]
Also Client API test failing - is this expected and if so make it explicit a change to the API is being made in the PR description.

File: console/src/cluster/notification.tsx [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+  const oldSever = status.data.oldServer as boolean;
```
Comment (pjdotson) [inline console/src/cluster/notification.tsx:21]: typo in `oldSever`

**Review by pjdotson — CHANGES_REQUESTED** [review]
